### PR TITLE
feat: #300 add useload hook

### PR DESCRIPTION
**Pull request by Howl404**

### CODE REVIEW

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [inline mw-webapp/src/hooks/useLoad.ts:6]: we can use real words in generics, T and D mean nothing in the current context. try to rename them, I hope it could increase readability a lot)

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [inline mw-webapp/src/hooks/useLoad.ts:31]: Should dependency be kind of react state? If yes, probably it will be better to have a list of dependencies, agree?

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [inline mw-webapp/src/hooks/useLoad.ts:63]: Could you describe the case when error is not instance of Error and why we should do nothing in this case?

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (tritonJS826) [inline mw-webapp/src/logic/userPage/UserPage.tsx:97]: do we still need to check "!userPreview"?

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (tritonJS826) [inline mw-webapp/src/logic/userPage/UserPage.tsx:93]: yeah! Looks beautiful!

**Review by tritonJS826 — COMMENTED** [review]
Why don't we use useLoad on  WayPage and UsersPage?

**Comment by Howl404** [comment]
@tritonJS826 Added only to two pages(one with validation and dependency and other with simple logic) to show example of usage, will implement on other pages after hook approve

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts:63]: I don't know much about other possible Errors, i think it can be "FirebaseError" too, don't really know how to handle everyting

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (Howl404) [reply mw-webapp/src/logic/userPage/UserPage.tsx:97]: Just for TS, so userPreview is not undefined in return

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts:31]: You mean something like `dependencies?: any[]`; and pass it to useEffect?

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts]: Renamed

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts]: Refactored

File: mw-webapp/src/hooks/useLoad.ts
Comment (Ekaterina1994) [inline mw-webapp/src/hooks/useLoad.ts:38]: Is it okey that hook useLoad not only fetch data but also valiadate it and other?

Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts:38]: We can rename it anytime, https://github.com/tritonJS826/masters-way/issues/300

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [inline mw-webapp/src/hooks/useLoad.ts:6]: Is it conventional for us to use the postfix type? Could it improve readability in current case?

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [reply mw-webapp/src/hooks/useLoad.ts:64]: - But catch block catches all errors in js by default. So why do you want to handle only one type and why do you want to ignore errors like "FirebaseErrors"

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (tritonJS826) [reply mw-webapp/src/logic/userPage/UserPage.tsx:97]: But if is loading finished and userPreview is not exist - this case mostly looks like a bug arequired to throw a error instead of showing loader?  It's not only about typescript as I see

Comment (Howl404) [reply mw-webapp/src/logic/userPage/UserPage.tsx:97]: This is impossible because we have onError callback on this page, i don't know how to skip userPreview check for render other way

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts:64]: I found that FirebaseError have instance of Error, so it will work

Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts:64]: We need to use instanceof to convert unknown into Error type, we can't pass it into onError function if we don't do it

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (Howl404) [reply mw-webapp/src/logic/userPage/UserPage.tsx:97]: Maked onError not optional

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts]: Removed postfix

**Review by Ekaterina1994 — APPROVED** [review]

**Review by SergioAJS — APPROVED** [review]

File: mw-webapp/src/logic/allWaysPage/AllWaysPage.tsx
Comment (tritonJS826) [inline mw-webapp/src/logic/allWaysPage/AllWaysPage.tsx:41]: Could we make dependencies optional?

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [reply mw-webapp/src/hooks/useLoad.ts:63]: So I still don't understand why should not you handle errors if they have other type, could you explain it please?

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [inline mw-webapp/src/hooks/useLoad.ts:37]: Should we use constants for consistency?

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [inline mw-webapp/src/hooks/useLoad.ts:31]: Dependencies->Dependency

**Review by tritonJS826 — COMMENTED** [review]
That's good, I will give an approve after implementation on all pages!

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (tritonJS826) [reply mw-webapp/src/logic/userPage/UserPage.tsx:97]: But it looks a little strange semantically for my opinion. When I see this line I think that there are two possible conditions when data is in "loading" state. Should we check all states the same way? If yes - why do we need isLoading state?

File: mw-webapp/src/hooks/useLoad.ts
Comment (Howl404) [reply mw-webapp/src/hooks/useLoad.ts:63]: @tritonJS826 Any error that can occur have instance of Error, because they are all based on it, FirebaseError = Error + additional properties

File: mw-webapp/src/logic/userPage/UserPage.tsx
Comment (Howl404) [reply mw-webapp/src/logic/userPage/UserPage.tsx]: Removed isLoading from this page

File: mw-webapp/src/hooks/useLoad.ts
Comment (tritonJS826) [reply mw-webapp/src/hooks/useLoad.ts:63]: ok, if it is really safe

**Review by tritonJS826 — APPROVED** [review]

**Pull request merged**
